File: Data Structures & Algorithms/count-number-of-islands/submission-2.py
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        num_of_islands = 0
        rows, cols = len(grid), len(grid[0])
        DIRECTIONS = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def bfs(row, col):
            queue = deque()
            queue.append((row, col))
            grid[row][col] = '#'

            while queue:
                curr_row, curr_col = queue.popleft()

                for dr, dc in DIRECTIONS:
                    new_row = curr_row + dr
                    new_col = curr_col + dc

                    if new_row < 0 or new_col < 0 or new_row >= rows or new_col >= cols or grid[new_row][new_col] != '1':
                        continue
                    queue.append((new_row, new_col))
                    grid[new_row][new_col] = '#'


        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == '1':
                    bfs(row, col)
                    num_of_islands += 1

        return num_of_islands

File: Data Structures & Algorithms/rotting-fruit/submission-2.cpp
class Solution {
private:
    using coords = pair<int, int>;
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int min_minutes = 0;
        int fresh_oranges = 0;

        queue<coords> q;
        vector<vector<int>> directions = {{0,1}, {0,-1}, {1, 0}, {-1,0}};

        for (int row = 0; row < grid.size(); row++) {
            for (int col = 0; col < grid[0].size(); col++) {
                if (grid[row][col] == 1) {
                    fresh_oranges += 1;
                } else if (grid[row][col] == 2) {
                    q.push({row, col});
                }
            }
        }

        while (!q.empty() && fresh_oranges > 0) {
            min_minutes += 1;

            int queue_size = q.size();
            for (int j = 0; j < queue_size; j++) {
                coords data = q.front();
                int curr_row = data.first;
                int curr_col = data.second;

                q.pop();

                for (int k = 0; k < directions.size(); k++) {
                    int new_row = curr_row + directions[k][0];
                    int new_col = curr_col + directions[k][1];

                    if ((0 <= new_row && new_row < grid.size()) &&
                        (0 <= new_col && new_col < grid[0].size()) &&
                        (grid[new_row][new_col] == 1)) {
                            grid[new_row][new_col] = 2;
                            fresh_oranges -= 1;
                            q.push({new_row, new_col});
                        }
                }
            }
        }

        return (fresh_oranges == 0) ? min_minutes : -1;
        
    }
};
